lib/at_cmd_parser/src/at_cmd_parser.c

```diff
@@ -0,0 +1,213 @@
+
+		if ((value < check_value) || (value == UINT32_MAX)) {
+			return -EINVAL;
+		}
+		at_str++;
+		(*consumed)++;
+	}
+
+	if (negative) {
+		value = (u32_t)(0 - value);
+	}
+
+	*val = value;
+	return 0;
+}
+
+
+/* Internal function. Parameters cannot be null. String must be null terminated.
+ */
+static int at_parse_param_numeric(const char *at_params_str,
+				       struct at_param_list *const list,
+				       size_t index, size_t *consumed)
+{
+	char *at_str = (char *)at_params_str;
+	size_t num_spaces = at_params_remove_spaces_from_beginning(&at_str);
+
+	u32_t val;
+	size_t consumed_bytes;
+	u32_t err = at_parse_param_u32(at_str, &val, &consumed_bytes);
+
+	if (err) {
+		*consumed = 0;
+		return -EINVAL;
+	}
+
+	if (val <= USHRT_MAX) {
+		err = at_params_put_short(list, index, (u16_t)(val));
+	} else {
+		err = at_params_put_int(list, index, val);
+	}
+
+	*consumed = consumed_bytes + num_spaces;
+	return 0;
+}
+
+
+/* Internal function. Parameters cannot be null. String must be null terminated.
+ */
+static int at_parse_param_string(const char *const at_params_str,
+				      struct at_param_list *const list,
+				      size_t index, size_t *consumed)
+{
+	u16_t str_len;
+	bool in_double_quotes = false;
+	char *str = (char *)at_params_str;
+	char *param_value_start = 0;
+
+	if (str == NULL || *str == '\0') {
+		return 0;
+	}
+
+	size_t spaces = at_params_remove_spaces_from_beginning(&str);
+
+	if (*str == '\"') {
+		str++;
+		in_double_quotes = true;
+	}
+
+	param_value_start = str;
+
+	while ((*str != '\0') && ((!in_double_quotes
+			&& ((*str != AT_CMD_SEPARATOR)
+			|| (*str != AT_CMD_PARAM_SEPARATOR)))
+			|| (in_double_quotes && (*str != '\"')))) {
+		str++;
+	}
+
+	if (in_double_quotes &&
+	    ((*str == AT_CMD_SEPARATOR) || (*str == '\0'))) {
+		return -EINVAL;
+	}
+
+	str_len = str - param_value_start;
+	*consumed = str_len + spaces;
+
+	if (in_double_quotes && (*str == '\"')) {
+		str++;
+		*consumed += 2;
+	}
+
+	return at_params_put_string(list, index, param_value_start, str_len);
+}
+
+
+/* Internal function. Parameters cannot be null. String must be null terminated.
+ */
+static int at_parse_param(const char *const at_params_str,
+			       struct at_param_list *const list, size_t index,
+			       size_t *consumed)
+{
+	if (at_params_str == NULL || *at_params_str == '\0' ||
+	    *at_params_str == AT_CMD_PARAM_SEPARATOR ||
+	    *at_params_str == AT_CMD_SEPARATOR) {
+		*consumed = 0;
+		(void)at_params_clear(list, index);
+		return 0;
+	}
+
+	int err = at_parse_param_numeric(at_params_str, list, index,
+					 consumed);
+
+	if (err) {
+		err = at_parse_param_string(at_params_str, list, index,
+					    consumed);
+	}
+
+	return err;
+}
+
+
+int at_parser_params_from_str(const char *const at_params_str,
+			      struct at_param_list *const list)
+{
+	return at_parser_max_params_from_str(at_params_str, list,
+					     list->param_count);
+}
+
+
+int at_parser_max_params_from_str(const char *const at_params_str,
```

Comment:
You declare `at_params_str` as constant pointer, then copy it and cast it to a non-constant one to manipulate it. I suggest to accept a non-constant pointer parameter to begin with, and get rid of the copy-casting.

Reply:
This is by design. You are not allowed to modify the input String. This is why we use a local variable.